### ISSUE TITLE
Setting Dictionaries

### ISSUE THREAD

**Issue by bretcope** [NFig/NFig#15]
NFig is intended to be statically typed. It replaces error-prone interfaces such as `Config.Get("Setting.Name")` with a nice statically named and typed interface `Config.Setting.Name`. However, I've been presented with a few use cases where something in-between static and dynamic would be useful.

For example, imagine your application has a list of scheduled tasks, and you'd like to setup an enabled/disabled flag for each. It's annoying to have to create an explicit settings property every time you create a new task, _and_ have to make sure each task is checking against the correct setting.

## Proposal

I'm proposing a "Setting Dictionary". Its usage would look something like this:

```csharp
public class Settings : NFigSettingsBase<SubApp, Tier, DataCenter>
{
    [Setting(true)]
    public SettingDictionary<bool> ScheduledTasksEnabled { get; }
}
```

`SettingDictionary<T>` would implement `IReadOnlyDictionary<string, T>` so you could use it like this:

```csharp
abstract class ScheduledTask
{
    public abstract string Name { get; }
    public bool IsEnabled => Config.Settings.ScheduledTasksEnabled[Name];
}
```

> I prefer implementing just `IReadOnlyDictionary` instead of `IDictionary` because I don't want to imply adding/editing/deleting a value will actually affect an override.

The dictionary's key is always a string, but the value can be any type that NFig supports (or any type you want via a custom converter).

When using the indexer syntax, if the key doesn't exist, the default value provided in the setting attribute will be returned (`true` in this case). I'm open to debate, but I think this behavior matches the common use case better than throwing an exception would. You can use `.ContainsKey(x)` to check whether a key actually exists. `.TryGetValue(x, out y)` will also return false if the key does not exist.

All of the attributes which are applicable to regular settings will be applicable to SettingDictionary settings. You can provide different default values for different sub-apps/tiers/data centers. You can provide a custom converter. You can encrypt them. The only thing to realize is that the attributes apply to each _individual value_ of the dictionary, rather than to the dictionary as a top-level thing.

### Known Keys

Obviously the admin panel will need to be able to set and clear arbitrary keys in a SettingDictionary. However, in some (most?) cases, the keys that the app cares about may actually be _known_ at runtime. Providing these known keys to the admin panel would allow for a better UX.

I'm not sure what the best way to provide these would be. In an ideal world, we could get the list of known keys when NFig is initialized. However, I'm worried that will cause a cyclical dependency for some people. Imagine trying to initialize settings, which wants to know the list of scheduled tasks, but scheduled tasks need settings to be initialized first.

The downside of providing known keys after initialization is that metadata would have to be regenerated. Perhaps known keys can be separated out into its own metadata somehow. Or maybe we just live with regenerating the metadata. I'm open to ideas on the matter. Either way, the core proposal doesn't need to be blocked on this feature.

There is also a question of whether `ContainsKey()` should return true if the key is known, even if no override is set. I'm leaning pretty strongly towards no. I think it's simpler to have "known keys" be strictly a metadata concept.

**Comment by rossipedia** [comment]
I love this. But definitely want to dig more into the UI/UX of the admin panel. By default I'd just have a dropdown of the keys in the dict, but I think it can be better.
